src/AudioToolbox/MusicSequence.cs

```diff
@@ -274,6 +288,39 @@ public double GetBeatsForSeconds (double seconds)
 				return beats;
 			return 0;
 		}
+
+		[DllImport (Constants.AudioToolboxLibrary)]
+		extern static /* OSStatus */ MusicPlayerStatus MusicSequenceSetUserCallback (/* MusicSequence */ IntPtr inSequence, MusicSequenceUserCallbackProxy inCallback, /* void * */ IntPtr inClientData);
+
+		public void SetUserCallback (MusicSequenceUserCallback callback)
+		{
+			lock (userCallbackHandles) {
+				if (userCallbackHandles.ContainsKey (handle))
+					userCallbackHandles[handle] = callback;
+				else
+					userCallbackHandles.Add (handle, callback);
+			}
+
+			MusicSequenceSetUserCallback (handle, userCallbackProxy, IntPtr.Zero);
+		}
+
+#if !MONOMAC
+		[MonoPInvokeCallback (typeof (MusicSequenceUserCallbackProxy))]
+#endif
+		static void UserCallbackProxy (IntPtr inClientData, IntPtr inSequence, IntPtr inTrack, double inEventTime, IntPtr inEventData, double inStartSliceBeat, double inEndSliceBeat)
+		{
+			MusicSequenceUserCallback userCallback;
+			lock (userCallbackHandles)
+				userCallbackHandles.TryGetValue (inSequence, out userCallback);
+
+			if (userCallback != null) {
+				var userEventData = new MusicEventUserData (inEventData);
+				var musicSequence = MusicSequence.Lookup (inSequence);
+				var musicTrack = new MusicTrack (musicSequence, inTrack, false);
+
+				userCallback (musicTrack, inEventTime, userEventData, inStartSliceBeat, inEndSliceBeat);
```

Comment:
Question: do we know (sample?) how much time the callback can be called ? because this allocates memory each time it's called (and let the GC deal with it) so if it's called a lot, under a short amount of time, you could run out of memory and/or and less predictable GC pauses.

In such case using `Runtime.GetINativeObject` could help, as it would reuse an existing instance (e.g. `MusicTrack`) but maybe it's better to always dispose after the call.